articles/ml_system_design/real_time_news_search_using_upstash_kafka_and_vector/upstash_ingest/upstash_ingest/ui.py

```diff
@@ -0,0 +1,98 @@
+"""
+    Streamlit UI for querying the Upstash Vector.
+    Structure:
+        - A text field
+        - A container section with columns for displaying search results
+        - Each search result is a card with an image, title, date, and a button to see more
+        - The button opens the original article in a new tab
+    Note:
+        It might take a few seconds to load the images, because they are downloaded from the web.
+"""
+
+from io import BytesIO
+
+import requests
+import streamlit as st
+from embeddings import TextEmbedder
+from PIL import Image
+from settings import settings
+from upstash_vector import Index
+
+v_index = Index(url=settings.UPSTASH_VECTOR_ENDPOINT, token=settings.UPSTASH_VECTOR_KEY)
+
+st.title("Upstash Real-Time News Search")
+results_placeholder = st.empty()
+
+
+def download_and_resize_image(url):
+    try:
+        response = requests.get(url)
+        if response.status_code == 200:
+            img = Image.open(BytesIO(response.content))
+            resized_img = img.resize((200, 300))
+            return resized_img
+        else:
+            st.error("Failed to download image.")
+    except Exception as e:
+        st.error(f"Error downloading image: {e}")
+
+
+def query_index(question: str):
+    embedder = TextEmbedder()
+    embds = embedder(question, to_list=True)
```

Comment:
What happens if the question is dirty or exceeds the context window of the embedding model?

Reply:
Also, I would rename the directory from `upstash_ingest` to `upstash_vector_db_ingest` or something like that.

Reply:
Great remark, forgot about it.
Will add cleaners(question) before actual embds.